content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1088 @@
+execve("./fork_exec", ["./fork_exec"], 0x7ffdb2c85028 /* 60 vars */) = 0
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 119051 attached
+[...]
+[pid 119051] execve("/home/student/.local/bin/ls", ["-la"], 0x7ffee90fee78 /* 60 vars */) = -1
+[...]
+[pid 119051] execve("/usr/bin/ls", ["-la"], 0x7ffee90fee78 /* 60 vars */) = 0
+```
+
+* The "manual" method (`fork` + `exec`) tries all `$PATH` entries before finding `"/usr/bin/ls"`
+
+----
+
+### Libc vs Linux vs Windows
+
+* Windows approach: `CreateProcess`
+
+```
+PS C:\...\lecture\demo\create-process> .\create_process.exe
+[parent] PID = 11876; child PID = 10376; Waiting...
+[...]
+```
+
+* The command given to `CreateProcess` must specify the interpreter: `cmd`
+
+---
+
+### `fork` vs `pthread_create`
+
+``` [1 - 2 | 3 - 4]
+student@os:~/.../compute/lecture/demo/create-process$ strace -e clone ./fork_exec
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7e83aa4810) = 5302
+student@os:~/.../compute/lecture/demo/create-thread$ strace -e clone ./create_thread
+clone(child_stack=0x7f9ea7df0fb0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, tls=0x7f9ea7df1700, child_tidptr=0x7f9ea7df19d0) = 5389
+```
+* From the flags, the parent and the child **thread** share:
+    * `CLONE_VM`: VAS (no copy-on-write)
+    * `CLONE_FS`: FS information (root, umask, cwd)
+    * `CLONE_FILES`: File descriptor table
+    * `CLONE_SIGHAND`: Signal handlers
+
+---
+
+### The Suspended States
+
+![Extended Thread Diagram](./media/thread-states-extended.svg)
+
+----
+
+### The Suspended States
+
+* **SUSPENDED READY / WAITING:** placed on secondary storage (**swap area**) due to lack of RAM
+    * Brought back to READY / WAITING when more RAM is available
+* Threads can move from SUSPENDED WAITING to SUSPENDED READY when I/O completes
+* Impossible to move from SUSPENDED READY to SUSPENDED WAITING (threads cannot perform I/O calls without running)
+
+---
+
+### Kernel Level Threads
+
+* Advantages:
+    * blocking actions only stall current thread
+    * hardware (SMP) support
+* Disadvantages:
+    * slow context switch
+    * difficult to implement and maintain
+    * kernel space scheduler (more complex)
+    * Partial solution: **user level threads**
+
+----
+
+### Thread Implementations
+
+| User level                       | Kernel level                               |
+| :------------------------------: | :----------------------------------------: |
+| implemented by libraries         | implemented by the kernel                  |
+| blocking actions stall process   | blocking actions only stall current thread |
+| more mapped on one kernel thread | provide support for user level threads     |
+| Java, Python, POSIX threads      | Linux KThreads, Windows threads            |
```

Comment:
**POSIX threads** are not guaranteed to be User-level, it depends on the implementation.